### DISCUSSION TITLE
Carbon error: Data object from model with relationship

### DISCUSSION THREAD

**Discussion by alexrififi** [spatie/laravel-data#91] [Q&A] [answered]
Imagine the following structure:

App has `clients` and `users` tables.
`Client` has many users and `User` has one `Client`
Both `Client` and `User` models has Data object, here they are:

```php
class ClientData extends Data
{
    public function __construct(
        public ?int $id,
        public ?Carbon $created_at,
        public ?Carbon $updated_at,
    )
    {
    }
}
```

```php
class UserData extends Data
{
    public function __construct(
        public int $id,
        public ?ClientData $client,
        public ?Carbon $created_at,
        public ?Carbon $updated_at,
    )
    {
    }
}
```

It is necessary to display a collection of `Users`, inside each `User` object there must be a `Client` related to it.
The task seems to be simple!

```php
UserData::collection(User::with('client')->paginate());
```

But, I got error: 
> Could not cast date: `2022-02-07T10:49:26.000000Z` into a `Illuminate\\Support\\Carbon` using format `Y-m-d\\TH:i:sP`

Here's why it's happening: 
1. In DataFromModelResolver model [transform to array](https://github.com/spatie/laravel-data/blob/main/src/Resolvers/DataFromModelResolver.php#L16)
2. Then datetime attributes transform into Carbon instance, but only model attributes, not relation model attributes
3. After executing the `DataFromModelResolver::execute` method, the result looks like this:
```php
[
  ...
  "created_at" => Illuminate\Support\Carbon
  "client" => [
    ...
    "created_at" => "2022-02-07T10:49:26.000000Z"
  ]
]
```

As you can see, the model attribute `created_at` and the same nested model attribute `created_at` have different values 

That is the problem. 
Tell me good people - how to win?

**Comment by wolkenheim** [comment, marked as answer]
Same here. I spent a few hours with this. I assume that nested relations are somehow normalized to arrays. So that means Carbon objects in Laravel models are serialized to strings with Carbon´s toIsoString() method. Does this help?
`#[WithCast(DateTimeInterfaceCast::class, format: "Y-m-d\TH:i:s.u\Z")]`

**Reply by alexrififi** [reply]
Yes, it helped! Thanks buddy!
But within the this package, it looks like a kind of crutch ...

Especially considering that when converting a model to a Date object, there is a [special process with dates](https://github.com/spatie/laravel-data/blob/main/src/Resolvers/DataFromModelResolver.php#L18-L20)

**Reply by wolkenheim** [reply]
Cool. There is a second option here. Not using any casts but writing dedicated magic fromModel() methods for every model (I used this option for other reasons). When using this option no casting is needed because you´re mapping Carbon object (from the model) to Carbon object (on the Data object). And this works. 

In the first case reflection for nested relations seem not to pass in models but arrays. That means Laravels default serialisation kicks in first as I described. That means there is no model passed as payload but already an array. So the DataFromModelResolver is not used for the children but the DataFromArrayResolver (if I´m wrong here correct me). 

Inside the data mapping process when no casting attribute is found, the global DateTimeInterface is called with the global format specified in config/data.php. The error is simply caused that this mapper expects the serialized date string in a different format.
You could specify the format  `  'date_format' => "Y-m-d\TH:i:s.u\Z", ` here and work without the custom casting. Only then you´re bound to stick with this for casting and transforming.  "Y-m-d\TH:i:s.u\Z" is the format toIsoString() uses and I spent literally hours to find that

**Reply by alexrififi** [reply]
> There is a second option here. Not using any casts but writing dedicated magic fromModel() methods for every model (I used this option for other reasons). When using this option no casting is needed because you´re mapping Carbon object (from the model) to Carbon object (on the Data object). And this works.

This was the first thing I tried, but it seemed not very convenient, I would like it to be resolved magically and I went here to discuss it  😄
```php
    public static function fromModel(User $user): self
    {
        return new self(
            $user->id,
            ...
            $user->created_at,
            $user->updated_at,
            Lazy::create(static fn() => $user->client?->getData()),
        );
    }
```


> In the first case reflection for nested relations seem not to pass in models but arrays. That means Laravels default serialisation kicks in first as I described. That means there is no model passed as payload but already an array. So the DataFromModelResolver is not used for the children but the DataFromArrayResolver (if I´m wrong here correct me).

Yes, I understand how it works and what caused the original error. I just wanted to draw the attention of developers to this nuance in that pure way 

> You could specify the format 'date_format' => "Y-m-d\TH:i:s.u\Z", here and work without the custom casting

Seems to be the best solution.
However, the output format changes: instead of  `2022-02-07T10:49:26+00:00` the output is `2022-02-07T10:49:26.000000Z`
But it's acceptable, big thanks for the tip!

**Comment by rubenvanassche** [comment]
Hi all,

This should be fixed with the release later today! https://github.com/spatie/laravel-data/commit/0c79a3b0588d8294245f4c0f30cdc65e756b297f

**Comment by moeen-basra** [comment]
apparently this is happening again.



```
{"error":{"code":400,"message":"Could not cast date `2023-02-22T23:50:33.000000Z` into a `Carbon\\CarbonImmutable` using formats: Y-m-d\\TH:i:sP"}}
```

**Reply by Slund** [reply]
Can confirm, had same issue recently.

**Comment by moeen-basra** [comment]
This work for me 

https://github.com/spatie/laravel-data/discussions/260#discussioncomment-4224429
